### PR TITLE
Allow users to control the TLS unwrapping behaviour.

### PR DESCRIPTION
**Pull request by Lukasa**
Motivation:

In some highly unusual protocol situations it is possible that TLS
unwrapping may want to time out without taking out the TCP connection.
This is most common when interoperating with peers that don't send
CLOSE_NOTIFY messages. It's a useful tool to be able to request that
timeouts don't take out the TCP connection.

Modifications:

- Added NIOSSLHandler.UnwrappingBehaviour.
- Added support for passing this to stopTLS.

Result:

Users can control unwrapping behaviour.

### CODE REVIEW

File: Sources/NIOSSL/NIOSSLHandler.swift [hunk 1]
```diff
@@ -552,6 +613,54 @@ extension NIOSSLHandler {
...
+    /// this alert, or reply very slowly. If the configured `TLSConfiguration.shutdownTimeout` expires before the remote
+    /// peer sends its reply `CLOSE_NOTIFY`, the `NIOSSLHandler` will give up. The `UnwrapTimeoutBehaviour` controls what
+    /// exactly is done in that "give up" state.
+    public struct UnwrapTimeoutBehaviour {
+        /// If the shutdown timeout expires, an error will be reported on the user promise, the `NIOSSLHandler` will
+        /// `close` the channel, and will also remove itself from the `ChannelPipeline`
+        public static let errorAndClose = UnwrapTimeoutBehaviour(.errorAndClose)
+
+        /// If the shutdown timeout expires, an error will be reported on the user promise, and the `NIOSSLHandler` will
+        /// remove itself from the `ChannelPipeline`, but it will not remove itself from the channel.
```
Comment (glbrntt) [inline Sources/NIOSSL/NIOSSLHandler.swift]: I don't understand removing the handler from the pipeline but not the channel; how does that work?

File: Tests/NIOSSLTests/UnwrappingTests.swift [hunk 2]
```diff
@@ -953,4 +957,177 @@ final class UnwrappingTests: XCTestCase {
...
+        for level in levels {
+            serverHandler.stopTLS(promise: promiseToAdd, onTimeout: level)
+            promiseToAdd = nil
+        }
+
+        XCTAssertFalse(clientClosed)
+        XCTAssertFalse(serverClosed)
+        XCTAssertFalse(unwrapped)
+
+        // Advance time by 5 seconds. This should fire the timeout. W
```
Comment (glbrntt) [inline Tests/NIOSSLTests/UnwrappingTests.swift]:
```suggestion
        // Advance time by 5 seconds. This should fire the timeout.
```

File: Sources/NIOSSL/NIOSSLHandler.swift (hunk 1 above)
Comment (Lukasa) [reply Sources/NIOSSL/NIOSSLHandler.swift]: It doesn't, that should say it will not _close_ the channel.

Comment (glbrntt) [reply Sources/NIOSSL/NIOSSLHandler.swift]: That makes much more sense; I should've made the connection

**Comment by Lukasa** [comment]
Thanks @glbrntt, both addressed.

**Comment by Lukasa** [comment]
Ok, this patch has been substantially rewritten. Instead of adding new API surface, we now just fire an error down the pipeline and don't close when the unwrap timeout fires. This gives users the same control with less API surface.

**Comment by glbrntt** [comment]
The approval is subject to the CI passing, of course 😃

**Comment by Lukasa** [comment]
@swift-nio-bot test this please

**Pull request merged**